### PR TITLE
service generator

### PR DESCRIPTION
**Pull request by slajax**
## overview

This aims to implement a generator for services as per the description in #4.

## note

depends on feathersjs/feathers-cli#41

## tasks

- [x] setup generator for use as `$ feathers generate service test`
- [x] allow generator to `--mount` onto [feathers-bootstrap](https://github.com/feathersjs/feathers-bootstrap)
- [x] support standalone vs generated apps with `--path` parameter
- [x] test out basic service generator prompts
- [x] generate static service config
- [x] configure static service hooks
- [x] configure static service filters
- [x] configure static service model
- [x] prompt prior to over writing paths
- [ ] integration tests

## technical debt
- prompts have been implemented but disabled due to complexity of workflow
- there are a few ares of technical debt that I noted with todos for refactor

## screenshots
screenshot 2016-10-15 14 58 02

### CODE REVIEW

**Comment by slajax** [comment]
@daffl @ekryski could use some feedback on mounting

**Comment by ekryski** [comment]
I think that `mount` option is good. What we might want to do is assume a default or allow the `mount` command to put it wherever you want as an override.

The other thing I have been working through is supporting whether it is a "standalone" service (ie. a microservice) or whether it is a service as part of an app. Right now we are assuming it is part of the same app but I guess to get around that, if you want to create a standalone microservice you easily just do `feathers g app` and then `feathers g service`

**Comment by slajax** [comment]
@ekryski yes that's exactly the intention, I believe I defaulted the option to false and early returned in the service generator if the default value was found or (need to add this actually) a file is not found at the defined `--mount` path.

---

Your point re: standalone service is actually a very important one. The first use case I was considering was `feather g app && feathers g service name` but I am also trying to consider the situation where someone didn't use the generator to organize their app because they followed the guides on the docs or wasn't aware of the generator when they started their project.

My thinking is that we can handle both:

```bash
$ mkdir app && cd app && mkdir src && mkdir src/services
$ feathers g service test --path src/services
$ vim src/server/app.js # manually implement service in feathers server
$ node src/server/app.js
```
_or_
```bash
$ feathers g app
$ feathers g service test --path server/services --mount server/feathers.json 
$ npm start
```

Once we settle on how this scales to filters, models etc we can figure out how to make `--path` and `--mount` self aware of what type of project they are in (via a config flag or directory searching) so generated projects can assume defaults and do not require these options to be passed to their generators eventually.

However, I think what you are you suggesting in your second point is that potentially there is a third use case where the service can be packaged and then mounted on a totally separate feathers application instance via fpm or something.

The key question seems like it is _should a standalone service export an entire (express mountable) service, or simply the service functions themselves_?

I guess my assumption for that use case has been you'll do something like:

```bash
$ mkdir test && cd test
$ feathers g service # maybe add --standalone but could probably be implicit
$ cd ../other-app/
$ feathers g app
$ feathers install feathersjs/test --path server/services --mount server/feathers.json
$ npm test
$ npm start
```

...maybe also support:
```bash
$ feathers link ../test --path server/services --mount server/feathers.json
```

...and rather then packaging the _entire_ feathers/express app, we would save some bloat by just exporting the functions from the generated assets. It would mean that we only mount services at install time in the second application instance, however I think the draw back here is that an exported app instance will be wayyyy easier to (headless) integration test via super test / request / http etc then it would be to unit test the raw functions even though in theory we could just initiate an app instance in the test harnesses which would allow us to make the express / feathers server dependencies dev only in services.

That said, I will chew on the pros and cons tomorrow and I'll also have to go look at feathers-bootstrap to double check that it supports mounting an exported app via the traditional `.configure()` or `.use()` express syntax.

Let me know if you think we're in alignment about where this is headed or if you have strong opinions about these design decisions. I'm not moving very fast on this, so I think we have plenty of time to hash it out as we go.

**Comment by slajax** [comment]
Another point to consider is if we export an entire application instance for services, it may end up as a one off because I wouldn't assume we'd do this for filters, models etc. Semantically, it might be better to have them follow the same approach for sanity sake.

File: src/service/middleware/rename.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export default function(options) {
+
+  return function mount(files, metalsmith, done){
+
+    debug(`Re-naming service templates to "${options.name}"`);
+
+    for(var k in files) {
+      files[k.replace('service', options.name)] = files[k];
+      delete files[k];
+    }
```
Comment (slajax) [inline src/service/middleware/rename.js]: so ugly, must refactor

File: src/service/templates/service.model.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+module.exports = function(options) {
+  // instantiate and return model adapter of choice
+}
```
Comment (slajax) [inline src/service/templates/service.model.js]: this should be generated from model templates, however I'm going to wait until we have traction on that as a standalone generator, less details get lost in the cross over

File: src/service/templates/service.json.js [hunk 3]
```diff
@@ -2,11 +2,13 @@ module.exports = function(config) {
   return {
     require: config.require,
     options: {
-      paginate: 'config.paginate',
-      Model: config.model
+      Model: `./${config.name}.model` ,
+      options: [config.model || {}]
```
Comment (slajax) [inline src/service/templates/service.json.js]: I'm thinking this should probably be tweaked so if you just pass an adapter type we can point it at a template / default instantiation of that model, but again, I'll give this some breathing room until we get better alignment with how the model generator works so we aren't duplicating templates

File: src/service/templates/service.model.js (hunk 2 above)
Comment (ekryski) [inline src/service/templates/service.model.js]: Agreed. We have some stuff that we need to talk to the @feathersjs/core-team about regarding these models.

**Comment by ekryski** [comment]
> However, I think what you are you suggesting in your second point is that potentially there is a third use case where the service can be packaged and then mounted on a totally separate feathers application instance via fpm or something.

Yes that was sort of what I was thinking. We essentially have 4 use cases:

1. You generate a service for the app that you are in the current working directory. A standard new generated app.
2. You generate a service for an app in a separate directory. A standard newly generated app.
3. You generate a standalone microservice (ie. an app + service). I'm thinking we just leave this as 2 steps like it is currently. You'd have to generate the app first and then the service you want to add.
4. You add a service to an existing app, non-generated app. This might be a bit trickier especially if the app is the old style and not using feathers-bootstrap. We don't have a good migration plan for old apps just yet so I'm open to ideas but also inclined to not worry about this for now.

So my inclination is to basically do what you are already doing 🎉 . If you generate a service we provide a mount point or assume it's the current working directory and a new feathers-bootstrap style app. If you want a standalone service, run `feathers g app` and then `feathers g service`.

**Comment by ekryski** [comment]
Let's :shipit:. We can improve tech debt and stuff in future PRs once we have the other generators in there. I want to play with this so that I can get my hooks and middleware changes compatible and push a PR.

**Comment by slajax** [comment]
Sounds good. I tested it a bunch in a vanilla generated app but it would be good to have eyes on it in a more mature area. I'll merge this and touch base with you tomorrow to see what you think comes next on this work stream. I assigned the tasks to you that I think you have traction on already.

**Pull request merged**